Implement MPGMRES - multiply preconditioned generalized minimal residual method

tamiko opened the pull request:
This is joint work with @wsmith-sch

We are interested in implementing the "multiply preconditioned" GMRES method (MPGMRES). This is a variant of the flexible GMRES method that uses multiple preconditioners to construct independent Krylov subspaces, one for each preconditioner. In contrast, the flexible GMRES  method implemented in SolverFGMRES constructs only one "Krylov subspace."

The changes required to our current FGMRES implementation are very minor. We will thus modify FGMRES to always do MPGMRES - and the special case of providing only one preconditioner object is FGMRES.

The design decision we have to make - and the reason for this early pull request - is the fact that we now need a `solve()` interface that can take more than one preconditioner as an argument. And here it gets a little bit awkward because we make no assumption on types (or existence of a base class). We suggest the following interface [1]:

```
SolverControl solver_control;
SolverMPGMRES<Vector<double>> solver(solver_control);

PreconditionIdentity prec_1;
PreconditionIdentity prec_2;
PreconditionIdentity prec_3;

const auto preconditioners = std::tie(prec_1, prec_2, prec_3);
solver.solve(system_matrix, solution, system_rhs, preconditioners);
```

The suggested interface also works well with const references:
```
const auto &ref_prec_1 = prec_1;
const auto &ref_prec_2 = prec_1;
const auto &ref_prec_3 = prec_1;
const auto ref_preconditioners = std::tie(ref_prec_1, ref_prec_2, ref_prec_3);
solver.solve(system_matrix, solution, system_rhs, ref_preconditioners);
```

We'll populate this pull request with our implementation in the coming days.

---

[1] Ideally we would have loved to support something like this:
```
solver.solve(system_matrix, solution, system_rhs, {prec_1, prec_2, prec_3});
```
But this is not possible due to constraints on the automatic type deduction for dependent function arguments.
We could in principle go the full variadic templates route and support something like this
```
solver.solve(system_matrix, solution, system_rhs, prec_1, prec_2, prec_3);
```
... but the amount of boilerplate code we need to write for this is annoying.
```
const auto preconditioners = std::tie(prec_1, prec_2, prec_3);
solver.solve(system_matrix, solution, system_rhs, preconditioners);
```

kronbichler reviewed (approved):
Great! I like this, I think it makes sense to go this route.

bangerth commented:
What is the issue with variadic templates? I actually sort of like the idea, because then one could omit listing any preconditioners at all, and that would map to the identity preconditioner if done right. 

I suspect that you got stuck on calling preconditioners of different types one after the other. I think this here would do it:
```
#include <functional>
#include <initializer_list>

template <typename... Preconditioners>
void solve (Preconditioners... preconditioners)
{
  for (const auto p : std::initializer_list<std::function<void()>>{ preconditioners... })
    p();
}


// preconditioner objects of different types
void f1();
auto f2 = [](){};
std::function<void()> f3;

int main()
{
  solve(f1,f2,f3);
}
```

bangerth reviewed (commented):
Did you mean to attach the implementation too, or is that still work in progress?

bangerth commented:
Cool idea, btw!

tamiko commented:
@bangerth
> What is the issue with variadic templates?
No concrete issue. Just a preference for what I suggested. We can also go with
```
solver.solve(system_matrix, solution, system_rhs, prec_1, prec_2, prec_3);
```

> Did you mean to attach the implementation too?

No, not yet.
We wanted to discuss our intent and interface first.

So what about @wsmith-sch and I will implement the interface with the function signature that takes preconditioner arguments directly? We'll probably invert the implementation in this case (implement `MPGMRES` and special case the `solve()` function for FGMRES).

tamiko commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -725,6 +730,53 @@ class SolverFGMRES : public SolverBase<VectorType>
+  template <typename MatrixType, typename... PreconditionerTypes>
+  DEAL_II_CXX20_REQUIRES(
+    (concepts::is_linear_operator_on<MatrixType, VectorType> &&
+     concepts::is_linear_operator_on<PreconditionerType, VectorType>))
```
Fixed.

bangerth commented:
I don't feel particularly strongly -- if you prefer one way over the other, that's cool with me. The inversion seems reasonable: One is a special case of any number, so deriving GMRES from MPGMRES sounds right.

tamiko commented:
@bangerth The sequence operator `,` is a bit more elegant:
```
([&](const auto &it){ it.vmult(dst, src); }(preconditioners), ...);
```

tamiko commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -2079,6 +2199,47 @@ void SolverFGMRES<VectorType>::solve(const MatrixType         &A,
+                                     const VectorType         &b,
+                                     const PreconditionerType &preconditioner)
+{
+  LogStream::Prefix prefix("FGMRES");
```
Do we still need this `LogStream::Prefix` object? If no, then we could get rid of `solver_internal` and implement everything in `MPGMRES::solve()` directly.

bangerth replied:
I've long thought that we should just do away with `LogStream` everywhere but in the test suite. But if you remove it here, you should probably remove its use in all of the solvers. (And not just remove the prefix, but actually remove writing into `deallog`.)

tamiko replied:
I agree. But let's tackle this in a future pull request.

bangerth commented:
Ah, you bet that I'll love the sequence operator :-) You can do away with the lambda, though, for example by using `std::apply`.

kronbichler commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -1986,16 +2055,68 @@ SolverFGMRES<VectorType>::SolverFGMRES(SolverControl        &cn,
+    // index:
+    //
+
+    static std::size_t current_index = 0;
```
Do we need to have a static variable, or can we define `current_index` just at the top of this function and increment it inside the function? The reason is that I would like to use static variables only where really necessary. But overall, this is a very neat implementation.

bangerth replied:
Yes, every `static` variable is not thread-safe unless guarded by a mutex.

tamiko replied:
This is now refactored.

Also please note that this PR is still marked as a draft. So please expect things to be preliminary and to change.

tamiko commented:
/rebuild

tamiko commented:
We will need to make our short test more robust with respect to rounding errors.
```
The following tests FAILED:
	3263 - lac/solver_mpgmres_01.debug (Failed)
Errors while running CTest
```

tamiko commented:
@kronbichler This would be ready now - pending a changelog entry.

tamiko commented:
@bangerth The `static` is long gone :smile:

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -2079,6 +2332,63 @@ void SolverFGMRES<VectorType>::solve(const MatrixType         &A,
+  : SolverMPGMRES<VectorType>(
+      cn,
+      mem,
+      typename SolverMPGMRES<VectorType>::AdditionalData{
+        data.max_basis_size,
+        true,
+        data.orthogonalization_strategy})
```
If you derive from the `AdditionalData` class of `MPSolverGMRES`, you can just pass `data` to the base class's constructor here.

tamiko replied:
I am not sure I want to because the boolean `use_truncated_mpgmres_strategy` does not make much sense *semantically* for `SolverFGMRES`. And there is no good way of hiding the parameter with public inheritance.

bangerth replied:
You can't "hide" the parameter, but aren't all possible values equivalent for the case of only one preconditioner?

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -2079,6 +2332,63 @@ void SolverFGMRES<VectorType>::solve(const MatrixType         &A,
+        true,
+        data.orthogonalization_strategy})
+{
+  this->indexing_strategy = SolverMPGMRES<VectorType>::IndexingStrategy::fgmres;
```
This is awkward. I *think* what you are doing is set a flag in the base class. I think a better design choice would be to add this as an argument to `solve_internal` so that you don't need the `protected` member variable in the base class.

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -687,28 +736,58 @@ class SolverFGMRES : public SolverBase<VectorType>
+    full_mpgmres,
+  };
+
+  /**
+   * The indexing strategy chosen for MPGMRES. This internal variable gets
+   * set by the constructor of SolverFGMRES and SolverMPGMRES.
+   */
+  IndexingStrategy indexing_strategy;
```
As mentioned below, I think this would work better if you make it an argument of the `solve_internal` function, rather than part of the state of the entire object.

kronbichler replied:
I agree, that could make sense. But we will need to make the `AdditionalData` protected to allow it to be called by the derived class `FGMRES`.

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -1963,39 +2128,126 @@ double SolverGMRES<VectorType>::criterion()
+    // We cycle through all preconditioners and call the one with a
+    // matching index:
+    //
+    std::size_t n = 0;
+
+    const auto call_current_preconditioner = [&](const auto &preconditioner) {
+      if (n++ == current_index)
+        preconditioner.vmult(dst, src);
+    };
+
+    // https://en.cppreference.com/w/cpp/language/fold
+    (call_current_preconditioner(preconditioners), ...);
```
On second thought, I understand why that doesn't work (because we need to determine at compile time what type `nth_element()` should return). I still find this too hard to read. Just wrap all of this into a lambda that you can call `call_nth_preconditioner(current_index, preconditioners...)` that simply puts a name to the whole block and wraps the variable `n` that is otherwise just a counting index.

kronbichler commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -630,30 +634,62 @@ class SolverGMRES : public SolverBase<VectorType>
 /**
- * Implementation of the Generalized minimal residual method with flexible
- * preconditioning (flexible GMRES or FGMRES).
+ * Implementation of the multiply preconditioned generalized minimal
```
```suggestion
 * Implementation of the multiple preconditioned generalized minimal
```

kronbichler commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -630,30 +634,62 @@ class SolverGMRES : public SolverBase<VectorType>
- * per iteration compared to GMRES, except one application of the
- * preconditioner less at each restart and at the end of solve().
+ * We implement two strategies, a "full" and a "truncated" MPGMRES version
+ * that differ in how they construct Krylov subspaces. The full MPGMRES
+ * version constructs a Krylov subspace for every possible combination of
+ * preconditioner application to the initial residual. For two
```
If I read this and the example below correctly, then the number of combinations grows exponentially with the iteration count. This is also what the paper states in their Sec 2.4. I think it would make sense to state this behavior and the possible implications on the size of notion of iteration and costs. Maybe you can just refer to the reference and the tables in that paper?

tamiko replied:
We have improved the `@note` in the class documentation that addresses this point.

kronbichler commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -1963,39 +2128,126 @@ double SolverGMRES<VectorType>::criterion()
+{
+  constexpr std::size_t n_preconditioners = sizeof...(PreconditionerTypes);
+
+  //
+  // With each invocation of preconditioner_vmult we increase the
+  // variable current_index by one modulo n_preconditioners.
+  //
```
This is more a question of style: We do usually not use empty commented lines before and after a comment (unless in very rare occasions of long functions), so I'd prefer if we could change this here and below to
```suggestion
  // With each invocation of preconditioner_vmult we increase the
  // variable current_index by one modulo n_preconditioners.
```
(This is not an important comment, though.)

tamiko replied:
You just haven't seen the light of proper commenting styles yet.
(I have updated the comments as you suggested.)

kronbichler reviewed (approved):
I like this, nice code. It was easier than expected, as we only have to select a preconditioner index according to the cycle index, and a previous vector from which to continue building the Krylov subspace. I have one question (not listed below): I did not fully understand the associated paper when quickly reading through it, so I might have missed that: Does it really work with our current code to solve the Hessenberg system (via Givens rotations) to avoid pecularities like linear dependence along some basis vectors, especially for the full MPGMRES case? I thought one would need to pick up some additional corrections in the form of a block Krylov method when doing that. In any case, it might be just my limited understanding, and I guess we might find out when working with this.

tamiko commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -678,6 +716,17 @@ class SolverFGMRES : public SolverBase<VectorType>
      */
     unsigned int max_basis_size;
 
+    /**
+     * If set to true (the default) a "truncated" search space is
+     * constructed consisting of the span of independent Krylov space
+     * associated with each preconditioner. If set to false, the full
+     * MPGMRES strategy for constructing the search space is used. This
+     * space consists of all possible combinations of iterative
+     * preconditioner applications; see the documentation of SolverMPGMRES
+     * for details.
+     */
+    bool use_truncated_mpgmres_strategy;
```
No, we don't state anywhere that we always use both strategies??

bangerth replied:
But in the class documentation you say something along the lines of "In the paper cited above, they consider *all* of the permutations of preconditioner products, whereas in this class, we *only* consider pure powers of each preconditioner". That suggests to me that the class implements only one strategy. If that's not true, at least that wasn't clear to me from reading the class documentation.

tamiko replied:
Maybe this was true for an older version of the documentation. The class documentation now talks about the "full" and "truncated" variants contrasting both.

wsmith-sch replied:
Yes, I think there was an earlier commit (which has been amended) that said something to this effect. In the very first draft of the code, we only implemented the truncated indexing strategy.

tamiko commented:
@kronbichler Yes this is a good point. As a follow-up it would be a nice project to revisit the Hessenberg matrix strategy and implement something more akin to what is discussed in the paper. We actually tested what happens when you repeat the preconditioner (which creates a huge number of linearly dependent vectors). The (re)orthogonalization that we perform in the `ArnoldiProcess` is what saves us.

tamiko commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -1963,39 +2128,126 @@ double SolverGMRES<VectorType>::criterion()
+    // We cycle through all preconditioners and call the one with a
+    // matching index:
+    //
+    std::size_t n = 0;
```
@bangerth We had that originally but MSVC could not handle the `constexpr` within the lambda (claiming that it wasn't a compile time constant). I'll add a test for this scenario.

tamiko commented:
@bangerth @kronbichler Thank you for the reviews. I have addressed almost all of your comments with two exceptions (making `SolverFGMRES::AdditionalData` a derived struct from `SolverMPGMRES::AdditionalData`, and the `if constexpr(...)` that we cannot add due to MSVC being difficult).

kronbichler reviewed (approved):
I agree with the changes.

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -630,30 +634,72 @@ class SolverGMRES : public SolverBase<VectorType>
+ * @f{align*}{
+ *   r, P_1r, P_2r, P_1AP_1r, P_2AP_2r, P_1AP_1AP_1r, P_2AP_2AP_2r, \ldots
+ * @f}
+ * For reference, FGMRES uses the following construction:
+ * @f{align*}{
+ *   r, P_1r, P_2P_1r, P_1P_2P_1r, \ldots
+ * @f}
```
Is this actually true? Does FGMRES not use any multiplication with $A$ at all?

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -664,12 +710,14 @@ class SolverFGMRES : public SolverBase<VectorType>
     /**
      * Constructor. By default, set the maximum basis size to 30.
      */
-    explicit AdditionalData(const unsigned int max_basis_size = 30,
+    explicit AdditionalData(const unsigned int max_basis_size         = 30,
+                            const bool use_truncated_mpgmres_strategy = true,
                             const LinearAlgebra::OrthogonalizationStrategy
                               orthogonalization_strategy =
                                 LinearAlgebra::OrthogonalizationStrategy::
                                   delayed_classical_gram_schmidt)
```
This change is not backward compatible. To be backward compatible, you need to add the `use_truncated_...` argument at the end.

bangerth replied:
If you do this, it would probably be useful to also put the new member variable at the bottom of the class.

tamiko replied:
This `AdditionalData` object is for the `MPGMRES` class which we introduce. So there is no issue with backwards compatibility.

Do you want to have the boolean to go last?

bangerth replied:
Let's just keep the same order between the two classes. So yes, let's put the `bool` last.

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -725,10 +804,96 @@ class SolverFGMRES : public SolverBase<VectorType>
+  /**
+   * Standardized data struct to pipe additional data to the solver.
+   */
+  struct AdditionalData
+  {
+    /**
+     * Constructor. By default, set the maximum basis size to 30.
+     */
+    explicit AdditionalData( //
+      const unsigned int max_basis_size = 30,
+      const LinearAlgebra::OrthogonalizationStrategy
+        orthogonalization_strategy = LinearAlgebra::OrthogonalizationStrategy::
+          delayed_classical_gram_schmidt)
+      : max_basis_size(max_basis_size)
+      , orthogonalization_strategy(orthogonalization_strategy)
+    {}
+
+    /**
+     * Maximum basis size.
+     */
+    unsigned int max_basis_size;
+
+    /**
+     * Strategy to orthogonalize vectors.
+     */
+    LinearAlgebra::OrthogonalizationStrategy orthogonalization_strategy;
+  };
```
I think this is still an open question. I recognize that the base class has one additional parameter, but perhaps there is a way to avoid re-implementing the class? Presumably, with just one preconditioner, the "truncated or not" question becomes moot anyway, and you could just be using the `AdditionalData` class of the base class?

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -725,10 +804,96 @@ class SolverFGMRES : public SolverBase<VectorType>
+  SolverFGMRES(SolverControl        &cn,
+               const AdditionalData &data = AdditionalData());
+
+  /**
+   * Solve the linear system $Ax=b$ for x.
+   */
+  template <typename MatrixType, typename... PreconditionerTypes>
+  DEAL_II_CXX20_REQUIRES(
+    (concepts::is_linear_operator_on<MatrixType, VectorType> &&
+     (concepts::is_linear_operator_on<PreconditionerTypes, VectorType> && ...)))
+  void solve(const MatrixType &A,
+             VectorType       &x,
+             const VectorType &b,
+             const PreconditionerTypes &...preconditioners);
```
This is also still a question.

bangerth commented on include/deal.II/lac/solver_gmres.h:
```diff
@@ -1986,16 +2158,112 @@ SolverFGMRES<VectorType>::SolverFGMRES(SolverControl        &cn,
+  // applying exactly one preconditioner with each function invocation to
+  // the vector src and storing the result in dst:
+
+  const auto preconditioner_vmult = [&](auto &dst, const auto &src) {
+    // We have no preconditioner that we could apply
+    if (n_preconditioners == 0)
+      {
+        dst = src;
+        return;
+      }
+
+    apply_nth_preconditioner(current_index, dst, src);
+    current_index = (current_index + 1) % n_preconditioners;
+  };
```
I think this would be easier to read:
```suggestion
  const auto preconditioner_vmult = [&](auto &dst, const auto &src) {
    // We have no preconditioner that we could apply
    if (n_preconditioners == 0)
      dst = src;
    else
      {
        apply_nth_preconditioner(current_index, dst, src);
        current_index = (current_index + 1) % n_preconditioners;
      }
  };
```

bangerth reviewed (changes requested):
I was about to hit the "merge" button, but I do still have a couple of comments about the `AdditionalData` class. But we're getting very close, and I'm sure this will be merged within the next couple of days.

tamiko commented:
@bangerth As discussed, I have updated the order of the `AdditionalData` fields to make them consistent and I have incorporated the stylistic change for the `if (...) { return; }` statement.

Let's see whether this still passes the testsuite :sweat_smile:
